src/org/sparkboard/server/server.clj

```diff
@@ -523,13 +390,213 @@
+   (fn [_ context payload {:keys [modal! action-id-suffix]}]
+     ; "Post an Update" button (user opens modal to respond to broadcast)
+     (let [firebase-path (str "/slack-broadcast/" action-id-suffix)
+           reply-ref (.push (fire-jvm/->ref (str firebase-path "/replies")))]
+       (fire-jvm/set-value reply-ref {:from-channel-id (-> payload :channel :id)})
+       (modal! (screens/team-broadcast-response
+                 (-> payload :message :blocks first :text :text) ; broadcast msg
+                 (str firebase-path "/replies/" (.getKey reply-ref))))))})
+
+(defn handle-slack-api-request [params handlers]
+  (let [params (cond-> params (:payload params) (update :payload json->clj))]
+    (log/trace "[slack-api]" params)
+    (if (:challenge params)
+      (ring.http/ok (:challenge params))
+      (do
+        (try-future
+          (let [[handler-id & args]
+                (cond (:event params)
+                      (let [event-type (:type (:event params))
+                            event (:event params)
+                            user-id (if (map? (:user event)) (:id (:user event)) (:user event))
+                            context (slack-context (:team_id params) user-id)]
+                        [(keyword "event" event-type) context event])
+
+                      (:payload params)
+                      (let [{:as payload
+                             :keys [user team]
+                             payload-type :type} (-> (:payload params)
+                                                     (update-in [:view :private_metadata] #(if (str/blank? %)
+                                                                                             {}
+                                                                                             (transit/read %))))
+                            context (slack-context (:id team) (:id user))]
+                        (case payload-type
+                          "shortcut"
+                          [(keyword payload-type (:callback_id payload "UNSET")) context payload]
+
+                          ("view_submission" "view_closed")
+                          [(keyword payload-type (-> payload :view :callback_id)) context payload]
+
+                          "block_actions"
+                          (let [[action-id action-id-suffix] (-> payload :actions first :action_id
+                                                                 (str/split (re-pattern screens/action-id-separator)))]
+                            [(keyword payload-type action-id)
+                             context
+                             payload
+                             {:action-id-suffix action-id-suffix
+                              :modal! (fn
+                                        ([blocks]
+                                         (log/info :modal! (hiccup/->blocks blocks))
+                                         (slack/web-api "views.open"
```

Comment:
It may make sense to factor out this fn

Reply:
I've removed this and added some view helpers in the slack ns, which rely on a new dynamic var: https://github.com/sparkboard/sparkboard/pull/45/commits/e629fa67a5954312cc4d250d8dd35df2e3507e5d#diff-4fb98734dd17bcb63fac2abb0966053cR97

also added a 2-arity to web-api for auto-filling the token.